components/calendar/src/roc.rs

```diff
@@ -0,0 +1,483 @@
+//! // `Date` type
+//! let date_iso = Date::try_new_iso_date(1970, 1, 2)
+//!     .expect("Failed to initialize ISO Date instance.");
+//! let date_roc = Date::new_from_iso(date_iso, Roc);
+//!
+//! // `DateTime` type
+//! let datetime_iso = DateTime::try_new_iso_datetime(1970, 1, 2, 13, 1, 0)
+//!     .expect("Failed to initialize ISO DateTime instance.");
+//! let datetime_roc = DateTime::new_from_iso(datetime_iso, Roc);
+//!
+//! // `Date` checks
+//! assert_eq!(date_roc.year().number, 59);
+//! assert_eq!(date_roc.month().ordinal, 1);
+//! assert_eq!(date_roc.day_of_month().0, 2);
+//!
+//! // `DateTime` checks
+//! assert_eq!(datetime_roc.date.year().number, 59);
+//! assert_eq!(datetime_roc.date.month().ordinal, 1);
+//! assert_eq!(datetime_roc.date.day_of_month().0, 2);
+//! assert_eq!(datetime_roc.time.hour.number(), 13);
+//! assert_eq!(datetime_roc.time.minute.number(), 1);
+//! assert_eq!(datetime_roc.time.second.number(), 0);
+//! ```
+
+use crate::{
+    calendar_arithmetic::ArithmeticDate,
+    iso::IsoDateInner,
+    types::{self, Era},
+    Calendar, CalendarError, Date, DateTime, Iso,
+};
+use tinystr::tinystr;
+
+/// Year of the beginning of the Taiwanese (ROC/Minguo) calendar.
+/// 1912 ISO = ROC 1
+const ROC_ERA_OFFSET: i32 = 1911;
+
+/// The Republic of China (ROC) Calendar
+///
+/// The [ROC calendar] is a solar calendar used in Taiwan and Penghu, as well as by overseas diaspora from
+/// those locations. Months and days are identical to the [`Gregorian`] calendar, while years are counted
+/// with 1912, the year of the establishment of the Republic of China, as year 1 of the ROC/Minguo/民国/民國 era.
+///
+/// [Republic of China calendar]: https://en.wikipedia.org/wiki/Republic_of_China_calendar
+///
+/// The Republic of China calendar should not be confused with the Chinese traditional lunar calendar
+/// (see [`Chinese`]).
+///
+/// # Era codes
+///
+/// This calendar supports two era codes: `"minguo"`, corresponding to years in the 民國 (minguo) era (CE year 1912 and
+/// after), and `"before minguo"`, corresponding to years before the 民國 (minguo) era (CE year 1911 and before).
+#[derive(Copy, Clone, Debug, Default)]
+#[allow(clippy::exhaustive_structs)] // this type is stable
+pub struct Roc;
+
+/// The inner date type used for representing [`Date`]s of [`Roc`]. See [`Date`] and [`Roc`] for more info.
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
+pub struct RocDateInner(IsoDateInner);
+
+impl Calendar for Roc {
+    type DateInner = RocDateInner;
+
+    fn date_from_codes(
+        &self,
+        era: crate::types::Era,
+        year: i32,
+        month_code: crate::types::MonthCode,
+        day: u8,
+    ) -> Result<Self::DateInner, crate::Error> {
+        let year = if era.0 == tinystr!(16, "minguo") {
+            if year <= 0 {
+                return Err(CalendarError::OutOfRange);
+            }
+            year + ROC_ERA_OFFSET
+        } else if era.0 == tinystr!(16, "before minguo") {
```

Comment:
Nit: Wrong era codes. They should be `roc` and `roc-inverse`; see https://github.com/unicode-org/cldr/pull/2665/files

Reply:
Done